src/pg_copy.c

```diff
@@ -0,0 +1,574 @@
+							  copyStatement->is_program,
+							  copyStatement->attlist,
+							  copyStatement->options);
+				
+	if (copyState->binary) 
+	{ 
+		elog(ERROR, "Copy in binary mode is not currently supported");
+	}
+
+	PG_TRY();
+	{
+		/* Lock all shards in shared mode */
+		shardIntervalList = SortList(shardIntervalList, CompareTasksByShardId);
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+		}
+
+		while (true)
+		{
+			MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+			nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+			MemoryContextSwitchTo(oldContext);
+		
+			if (!nextRowFound)
+			{
+				MemoryContextReset(tupleContext);
+				break;
+			}
+			
+			/* write the row to the shard */
+			if (columnNulls[partitionColumn->varattno]) 
+			{ 
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			} 
+			rightConst->constvalue = columnValues[partitionColumn->varattno];
+			prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+				
+			foreach(shardIntervalCell, prunedList)
+			{
+				ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+				ShardId shardId = shardInterval->id;
+				bool found = false;
+				
+				shardConnections = (ShardConnections*)hash_search(shardToConn, &shardInterval->id, HASH_ENTER, &found);
+				if (!found) 
+				{ 
+					InitializeShardConnections(copyStatement, shardConnections, shardId, transactionManager);
+				}
+				lineBuf = CopyGetLineBuf(copyState);
+				lineBuf->data[lineBuf->len++] = '\n'; 
+				/* There was already new line in the buffer, but it was truncated: 
+				 * no need to check available space */
+				
+				/* Replicate row to all shard placements */
+				for (i = 0; i < shardConnections->replicaCount; i++) 
+				{ 
+					if (PQputCopyData(shardConnections->conn[i], lineBuf->data, lineBuf->len) <= 0)
+					{
+						elog(ERROR, "Copy failed for shard %ld", (long)shardId);
+					}
+				}
+			}
+			MemoryContextReset(tupleContext);
+		}
+		/* Perform two phase commit in replicas */
+		shardConnectionsList = HTABToList(shardToConn);
+		failedShard = DoForAllShards(shardConnectionsList, PgCopyPrepareTransaction, relationName);
+	}
+	PG_CATCH(); /* do recovery */
+	{
+		EndCopyFrom(copyState);
+		heap_close(rel, AccessShareLock);
+		/* Rollback transactions */
+		shardConnectionsList = HTABToList(shardToConn);
+		DoForAllShards(shardConnectionsList, PgCopyAbortTransaction, relationName);
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	EndCopyFrom(copyState);
+	heap_close(rel, AccessShareLock);
+
+	/* Complete two phase commit */
+	DoForAllShards(shardConnectionsList, PgCopyEndTransaction, relationName);
```

Comment:
If I understand correctly, we would still commit the prepares that succeeded even if one of them failed right now.